kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaRestoreService.java

```diff
@@ -0,0 +1,160 @@
+      Executors.newSingleThreadScheduledExecutor(
+          new ThreadFactoryBuilder().setNameFormat("replica-restore-service-%d").build());
+  private final BlockingQueue<SnapshotMetadata> queue = new LinkedBlockingQueue<>();;
+  private final ReplicaMetadataStore replicaMetadataStore;
+  private final MeterRegistry meterRegistry;
+  private final Counter successfullyCreatedReplicas;
+  private final Counter failedReplicas;
+  private final Counter skippedReplicas;
+  private final Timer restoreTimer;
+  protected static final Logger LOG = LoggerFactory.getLogger(ReplicaCreationService.class);
+  public static String REPLICAS_CREATED = "replicas_created";
+  public static String REPLICAS_FAILED = "replicas_failed";
+  public static String REPLICAS_SKIPPED = "replicas_skipped";
+  public static String REPLICAS_RESTORE_TIMER = "replicas_restore_timer";
+
+  public ReplicaRestoreService(
+      ReplicaMetadataStore replicaMetadataStore,
+      MeterRegistry meterRegistry,
+      KaldbConfigs.ManagerConfig managerConfig) {
+    this.managerConfig = managerConfig;
+    this.replicaMetadataStore = replicaMetadataStore;
+
+    this.meterRegistry = meterRegistry;
+    this.skippedReplicas = meterRegistry.counter(REPLICAS_SKIPPED);
+    this.successfullyCreatedReplicas = meterRegistry.counter(REPLICAS_CREATED);
+    this.failedReplicas = meterRegistry.counter(REPLICAS_FAILED);
+    this.restoreTimer = meterRegistry.timer(REPLICAS_RESTORE_TIMER);
+  }
+
+  /** Queues snapshot restoration task to be run in the future */
+  @Override
+  protected void runOneIteration() {
+    if (pendingTask == null || pendingTask.getDelay(TimeUnit.SECONDS) <= 0) {
+      pendingTask =
+          executorService.schedule(
+              this::restoreQueuedSnapshots,
+              managerConfig.getEventAggregationSecs(),
+              TimeUnit.SECONDS);
+    } else {
+      LOG.info(
+          "Replica restore task already scheduled, will run in {} ms",
+          pendingTask.getDelay(TimeUnit.MILLISECONDS));
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting replica restore service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closing replica restore service");
+    executorService.shutdownNow();
+    LOG.info("Closed replica restore service");
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getReplicaRestoreServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  /** Queues snapshots to be restored in the future */
+  public synchronized void queueSnapshotsForRestoration(List<SnapshotMetadata> snapshotsToRestore)
+      throws Exception {
+    if (snapshotsToRestore.size()
+        >= managerConfig.getReplicaRestoreServiceConfig().getMaxReplicasPerRequest()) {
+      throw new SizeLimitExceededException();
```

Comment:
You probably want to add a message in the `SizeLimitExceededException` params here that says how many were requested, and what the limit actually is. That'll bubble up via your GRPC service and be returned to the user then given the current code.